### ISSUE TITLE
What about "read groups" for BaseRecalibrator in GATK?

### ISSUE THREAD

**Issue by bgrueda** [bgrueda/WES_LUAD#11]
As part of the preprocessing of the sample I use the BaseRecalibration tool from GATK. 

```
gatk BaseRecalibrator -I mds_ex.bam -R hg38.fasta  --known-sites hg38.dbsnp138.vcf  -O xBR.table
```
and I have this error:

`A USER ERROR has occurred: Number of read groups must be >= 1, but is 0`

and I am not entirely sure what is a *read group*.

I found that, 

> The recalibration system is read-group aware, meaning it uses @RG tags to partition the data by read group. This 
> allows it to perform the recalibration per read group, which reflects which library a read belongs to and what lane it was sequenced in on the flowcell.

Looking around the web I found another tool from Picard that is called `AddOrReplaceReadGroups`, 
but I don't know what to write in the arguments required, since it is information about the sequencing. 

So, I am stock and I would like to know how to solve the problem and if is it possible to solve and use this tool.

Danke

This is the file that I used for this: 
[mds_ex.bam](https://github.com/bgrueda/WES_LUAD/blob/main/data/mds_ex.bam?raw=true)

and I am already using *gatk (v.4.1.9.0)*

the accessory files were downloaded from Broad Institute resources for hg38.
https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38/v0?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false

**Comment by IsauraRReinhold** [comment]
I found the same issue in the GATK page, maybe the answer is there

https://gatk.broadinstitute.org/hc/en-us/articles/360036898312-BaseRecalibrator

**Comment by IsauraRReinhold** [comment]
In bioistars I found many possibilities to resolve your issue

https://www.biostars.org/p/115819/

**Comment by IsauraRReinhold** [comment]
In this other link there are other ideas

https://www.biostars.org/p/47487/

**Comment by AlexLlanesQuevedo** [comment]
On the "read groups" concept:

"There is no formal definition of what a 'read group' is, however in practice this term refers to a set of reads that are generated from a single run of a sequencing instrument.

In the simple case where a single library preparation derived from a single biological sample was run on a single lane of a flow cell, all the reads from that lane run belong to the same read group. When multiplexing is involved, then each subset of reads originating from a separate library run on that lane will constitute a separate read group.

Read groups are identified in the SAM/BAM /CRAM file by a number of tags that are defined in the official SAM specification. These tags, when assigned appropriately, allow us to differentiate not only samples, but also various technical features that are associated with artifacts."

From [ GATK Technical Documentation Glossary](https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups)

**Comment by IsauraRReinhold** [comment]
https://github.com/ekg/bamaddrg

This person propose another solution to adds read groups.

bamaddrg: adds read groups to input BAM files, streams BAM output on stdout.

This is intended for use "fixing up" RG tags on the fly so that they reflect
the source file from which the aligment originated from.  This allows the
"safe" merging of many files from many individuals into one stream, suitable
for input into downstream processing systems such as freebayes (a population
variant detector).

Usage:

To tag multiple files simultaneously, use as such:

    % bamaddrg -b file1.bam -s jill -r group.s1.1 \
               -b file2.bam -s jill -r group.s1.2 \
               -b file3.bam -s bill \
               -b file4.bam \
       | freebayes ...

This would add the sample name "jill" to records from file1.bam and file2.bam,
while specifying that specific read group identify alignments from each file.
Alignments from file3.bam would be tagged with sample name "bill" and read
group id "bill", and samples from file4.bam would be tagged with sample name
"file4.bam" and read group id "file4.bam."

**Comment by AlexLlanesQuevedo** [comment]
In case of not being able to solve the issue by any previous means, the GATK team recommends
"If you don't know what information to put in the read groups, you should ask whoever performed the sequencing or provided the BAM to give you the metadata you need"
[Errors about read group (RG) information](https://gatk.broadinstitute.org/hc/en-us/articles/360035532352)

**Comment by bgrueda** [comment]
Using the command 
`perl -e 'print "\@RG\tID:ga\tSM:hs\tLB:ga\tPL:Illumina\n"' > rg.txt`

we created a file with the correct format for the *Read group*, as follows: 
`@RG	ID:ga	SM:hs	LB:ga	PL:Illumina`

Then, I added this line to my sam file and turned to bam
I used this new bam file with the first command I wrote. 

`gatk BaseRecalibrator -I mds_ex.bam -R hg38.fasta  --known-sites hg38.dbsnp138.vcf  -O xBR.table`

and IT WORKS!
The table that I needed was created and this issue is closed.
